overlord/configstate/configcore/cloud.go

```diff
@@ -0,0 +1,90 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package configcore
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"os"
+
+	"github.com/snapcore/snapd/dirs"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/state"
+)
+
+func alreadySeeded(tr Conf) (bool, error) {
+	st := tr.State()
+	st.Lock()
+	defer st.Unlock()
+	var seeded bool
+	err := tr.State().Get("seeded", &seeded)
+	if err != nil && err != state.ErrNoState {
+		return false, err
+	}
+	return seeded, nil
+}
+
+func handleCloud(tr Conf) error {
+	// if we are during seeding try to capture cloud information
+	seeded, err := alreadySeeded(tr)
+	if err != nil {
+		return err
+	}
+	if seeded {
+		// already done
+		return nil
+	}
+
+	data, err := ioutil.ReadFile(dirs.CloudInstanceDataFile)
```

Comment:
As you mentioned in TODO comments, might want either a retry on this file check as cloud-init will write the instance-data.json file before it performs any of the other user-data provided commands/snap management operations etc.

Reply:
Ok, false alarm. After a followup discussion in this, I didn't know about the systemd dependency chain in snapd that ensured snapd runs after cloud-init.service is complete. Scott Moser pasted the 'proof' of this dependency chain which will guarantee that snapd runs after any datasource is intialized and instance.json is written. http://paste.ubuntu.com/p/FM79xKC4dw/.